File: paper/R/fig_cuprite_comparison.R
# fig_cuprite_comparison.R --- Cuprite 6-panel figure with USGS Rockwell ref
# Migration of scripts/fig_cuprite_comparison.py.
# Reads composite from prep_rasters_for_R.py + USGS Rockwell raster.
# Run: Rscript paper/R/fig_cuprite_comparison.R  (cwd = project root)

suppressPackageStartupMessages({
  library(ggplot2)
  library(patchwork)
  library(terra)
  library(tidyterra)
  library(sf)
  library(dplyr)
  library(tibble)
})

source("paper/R/theme_paper.R")
setup_paper_theme("springer")

S2_TIF  <- "data/maps/cuprite_composite.tif"
GT_TIF  <- "data/ground_truth/cuprite_ground_truth.tif"
OUT_PDF <- "figures/cuprite_comparison.pdf"

# ---- Load rasters ----

r <- rast(S2_TIF)
names(r) <- c("B2", "B3", "B4", "B7", "B8", "B8A", "B11", "B12")
ext_r <- ext(r)
AOI <- list(west = ext_r[1], east = ext_r[2],
            south = ext_r[3], north = ext_r[4])

gt <- rast(GT_TIF)
# Ensure same crs as composite
if (st_crs(gt)$epsg != 4326) gt <- project(gt, "EPSG:4326")

# ---- SR formulas locally-rediscovered at Cuprite (Tabla tab:cuprite_formulas) ----

eps <- 1e-6
clamp_lo <- function(x, lo) clamp(x, lower = lo, upper = Inf, values = TRUE)
b2 <- r$B2; b3 <- r$B3; b4 <- r$B4; b7 <- r$B7
b8 <- r$B8; b8a <- r$B8A; b11 <- r$B11; b12 <- r$B12

sr1 <- log(clamp_lo(b11, eps) / clamp_lo(b12, eps))  # Silicic
sr2 <- b11 - b12                                       # Adv. Argillic
sr3 <- tanh(b11) - b8a                                 # Argillic-Phyllic
sr4 <- ((b12 - b8) / clamp_lo(b7, eps))^2              # Propylitic

percentile_stretch <- function(x, p_lo = 2, p_hi = 98) {
  v <- values(x); v <- v[is.finite(v)]
  q <- quantile(v, probs = c(p_lo, p_hi) / 100, na.rm = TRUE)
  (x - q[1]) / max(q[2] - q[1], eps)
}

sr_stack <- rast(list(
  Silicic         = percentile_stretch(sr1),
  AdvArgillic     = percentile_stretch(sr2),
  ArgillicPhyllic = percentile_stretch(sr3),
  Propylitic      = percentile_stretch(sr4)
))
sr_stack <- clamp(sr_stack, 0, 1)
classmap <- which.max(sr_stack)
levels(classmap) <- data.frame(
  ID = 1:4,
  class = c("Silicic", "Adv. Argillic", "Argillic-Phyllic", "Propylitic")
)

clay_ratio <- b11 / clamp_lo(b12, eps)

# USGS GT raster — encode as factor matching same class labels
gt_factor <- gt
levels(gt_factor) <- data.frame(
  ID = 1:4,
  class = c("Silicic", "Adv. Argillic", "Argillic-Phyllic", "Propylitic")
)
gt_factor[gt_factor == 0] <- NA

# Vectorize GT for outline overlay
gt_polys <- as.polygons(gt_factor, dissolve = TRUE, values = TRUE, aggregate = TRUE)
gt_sf <- st_as_sf(gt_polys)

# ---- True / false colour ----

stretch_rgb <- function(b1, b2, b3, gamma = 0.85) {
  do_one <- function(x) {
    s <- percentile_stretch(x); s <- clamp(s, 0, 1); s^gamma
  }
  rast(list(R = do_one(b1), G = do_one(b2), B = do_one(b3)))
}
rgb_true <- stretch_rgb(b4, b3, b2)
rgb_swir <- stretch_rgb(b12, b11, b4)

# ---- Plot helpers ----

map_theme <- function() {
  theme(
    legend.position = c(0.02, 0.02),
    legend.justification = c(0, 0),
    legend.background = element_rect(fill = "white", colour = "grey80",
                                     linewidth = 0.2),
    legend.title = element_text(size = 6, face = "bold"),
    legend.text = element_text(size = 5.5),
    legend.key.size = unit(0.3, "cm"),
    plot.subtitle = element_text(face = "bold", size = 8.5,
                                 hjust = 0.5, margin = margin(b = 2)),
    axis.text = element_text(size = 6.5),
    axis.title = element_text(size = 7),
    plot.margin = margin(4, 4, 4, 4, "pt")
  )
}

style_axes <- function(subtitle) {
  list(
    scale_x_continuous(name = "Longitude (°W)",
                       limits = c(AOI$west, AOI$east),
                       breaks = seq(round(AOI$west, 2), round(AOI$east, 2), 0.05),
                       expand = expansion(0)),
    scale_y_continuous(name = "Latitude (°N)",
                       limits = c(AOI$south, AOI$north),
                       breaks = seq(round(AOI$south, 2), round(AOI$north, 2), 0.05),
                       expand = expansion(0)),
    labs(subtitle = subtitle),
    coord_sf(crs = 4326, expand = FALSE)
  )
}

# ---- Build 6 panels ----

p_a <- ggplot() +
  geom_spatraster_rgb(data = rgb_true, max_col_value = 1) +
  annotate("segment", x = AOI$west + 0.005,
           xend = AOI$west + 0.005 + 0.035,
           y = AOI$south + 0.006,
           yend = AOI$south + 0.006,
           colour = "black", linewidth = 0.9) +
  annotate("text", x = AOI$west + 0.0225,
           y = AOI$south + 0.011,
           label = "~3 km", colour = "black",
           fontface = "bold", size = 2.2) +
  style_axes("Sentinel-2 true colour (B4-B3-B2)") +
  map_theme()

p_b <- ggplot() +
  geom_spatraster_rgb(data = rgb_swir, max_col_value = 1) +
  style_axes("SWIR false colour (B12-B11-B4)") +
  map_theme()

p_c <- ggplot() +
  geom_spatraster(data = clay_ratio) +
  scale_fill_scico(palette = "vik", direction = -1,
                   limits = quantile(values(clay_ratio),
                                     c(0.02, 0.98), na.rm = TRUE),
                   oob = scales::squish, na.value = NA,
                   name = expression(B[11]/B[12]),
                   guide = guide_colorbar(barwidth = unit(2.5, "cm"),
                                          barheight = unit(0.25, "cm"))) +
  style_axes("Clay Ratio (B11/B12)") +
  map_theme() +
  theme(legend.position = "bottom",
        legend.direction = "horizontal",
        legend.title = element_text(size = 6, face = "plain", vjust = 0.85))

p_d <- ggplot() +
  geom_spatraster(data = classmap) +
  scale_fill_manual(values = pal_alteration, na.value = NA, guide = "none") +
  style_axes("SR classification (Cuprite formulas)") +
  map_theme()

p_e <- ggplot() +
  geom_spatraster(data = gt_factor) +
  scale_fill_manual(values = pal_alteration, na.value = NA, guide = "none") +
  style_axes("USGS Rockwell (2017) reference") +
  map_theme()

p_f <- ggplot() +
  geom_spatraster(data = classmap, alpha = 0.85) +
  scale_fill_manual(values = pal_alteration, na.value = NA, guide = "none") +
  geom_sf(data = gt_sf, fill = NA, colour = "black",
          linewidth = 0.3) +
  style_axes("SR classification + USGS outlines") +
  map_theme()

fig <- (p_a | p_b | p_c) / (p_d | p_e | p_f) +
  plot_annotation(tag_levels = "a", tag_suffix = ")") &
  theme(plot.tag = element_text(face = "bold", size = 9))

dir.create("figures", showWarnings = FALSE)
save_paper(fig, OUT_PDF, width_cm = 18.0, height_cm = 12.0)

File: scripts/fig_cuprite_comparison.py
#!/usr/bin/env python3
"""
Cuprite comparison figure (addresses R#1 critique #7 from the NRR reject).

Six-panel figure for Cuprite, Nevada (USA):
  (a) Sentinel-2 true colour
  (b) Sentinel-2 SWIR false colour
  (c) Clay Ratio (B11/B12)
  (d) SR per-pixel classification using the four locally-rediscovered Cuprite
      formulas (Table tab:cuprite_formulas in the manuscript)
  (e) USGS Rockwell (2017) ASTER-derived alteration map — the reference
      product the reviewer asked us to compare against
  (f) Panel (d) with USGS reference outlines overlaid for visual co-registration

Reads:
  - data/ground_truth/cuprite_ground_truth.tif  (USGS Rockwell raster, EPSG:4326)
Pulls from GEE:
  - Sentinel-2 SR median composite over the same AOI
"""

import ee
import numpy as np
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.patches import Patch
from matplotlib.colors import ListedColormap, BoundaryNorm
import rasterio
from rasterio.features import shapes
from shapely.geometry import shape as shp_shape
import geopandas as gpd
import urllib.request, tempfile
from pathlib import Path

ee.Initialize()

FIG_DIR = Path("figures")
GT_PATH = Path("data/ground_truth/cuprite_ground_truth.tif")
SCALE_M = 30

# USGS Rockwell class mapping (matches the manuscript's Cuprite groupings)
USGS_CLASS_NAMES = {1: "Silicic", 2: "Adv. Argillic", 3: "Argillic-Phyllic", 4: "Propylitic"}
USGS_CLASS_COLORS = {1: "#FF7F00", 2: "#E41A1C", 3: "#4DAF4A", 4: "#00AA00"}

# ---------- 1. Load USGS GT raster + AOI from its bounds ----------
print("Loading USGS Rockwell ground truth...")
with rasterio.open(GT_PATH) as src:
    gt = src.read(1)
    gt_bounds = src.bounds
    gt_transform = src.transform
    gt_crs = src.crs
print(f"  GT shape: {gt.shape}, bounds: {gt_bounds}, CRS: {gt_crs}")
print(f"  Classified pixels: {(gt > 0).sum()} / {gt.size}")

AOI = dict(west=gt_bounds.left, south=gt_bounds.bottom,
           east=gt_bounds.right, north=gt_bounds.top)
print(f"  AOI: {AOI}")

# ---------- 2. Download S2 composite from GEE for the same AOI ----------
print("Downloading Sentinel-2 composite from GEE...")
aoi = ee.Geometry.Rectangle([AOI["west"], AOI["south"], AOI["east"], AOI["north"]])
s2 = (
    ee.ImageCollection("COPERNICUS/S2_SR_HARMONIZED")
    .filterBounds(aoi)
    .filterDate("2024-01-01", "2024-06-01")
    .filter(ee.Filter.lt("CLOUDY_PIXEL_PERCENTAGE", 10))
    .select(["B2", "B3", "B4", "B7", "B8", "B8A", "B11", "B12"])
)
print(f"  S2 scenes available: {s2.size().getInfo()}")
composite = s2.median().divide(10000)

url = composite.getDownloadURL({
    "scale": SCALE_M,
    "region": aoi,
    "format": "GEO_TIFF",
    "crs": "EPSG:4326",
})
print(f"  Downloading GeoTIFF...")
with urllib.request.urlopen(url, timeout=180) as resp:
    data = resp.read()
print(f"  Downloaded {len(data)/1e6:.1f} MB")

with tempfile.NamedTemporaryFile(suffix=".tif", delete=False) as tmp:
    tmp.write(data)
    tmp_path = tmp.name

BAND_ORDER = ["B2", "B3", "B4", "B7", "B8", "B8A", "B11", "B12"]
with rasterio.open(tmp_path) as src:
    arr = src.read()
    s2_bounds = src.bounds
print(f"  S2 raster shape: {arr.shape}, bounds: {s2_bounds}")
bands = {name: arr[i].astype(np.float32) for i, name in enumerate(BAND_ORDER)}

# ---------- 3. Compute Cuprite SR formulas (Table tab:cuprite_formulas) ----------
eps = 1e-6
sr = {
    1: np.log(np.maximum(bands["B11"], eps) / np.maximum(bands["B12"], eps)),  # Silicic
    2: bands["B11"] - bands["B12"],                                            # Adv. Argillic
    3: np.tanh(bands["B11"]) - bands["B8A"],                                   # Argillic-Phyllic
    4: ((bands["B12"] - bands["B8"]) / np.maximum(bands["B7"], eps)) ** 2,     # Propylitic
}
# Normalize each to [0,1] via p2/p98 percentile stretch, then argmax
sr_norm = {}
for k, v in sr.items():
    p2, p98 = np.nanpercentile(v[np.isfinite(v)], [2, 98])
    sr_norm[k] = np.clip((v - p2) / max(p98 - p2, eps), 0, 1)
sr_stack = np.stack([sr_norm[k] for k in sorted(sr_norm)], axis=0)
classmap = np.argmax(sr_stack, axis=0) + 1  # 1..4

clay_ratio = bands["B11"] / np.maximum(bands["B12"], eps)

# ---------- 4. Vectorize USGS GT for the outline overlay panel ----------
print("Vectorizing USGS classes for outline overlay...")
gt_polys = []
for cls_id in range(1, 5):
    mask = (gt == cls_id).astype("uint8")
    for geom, val in shapes(mask, mask=mask.astype(bool), transform=gt_transform):
        if val == 1:
            gt_polys.append({"class_id": cls_id, "geometry": shp_shape(geom)})
gdf = gpd.GeoDataFrame(gt_polys, crs=gt_crs)
# Simplify a bit so overlay edges are clean
gdf["geometry"] = gdf["geometry"].simplify(0.0003)
print(f"  GT polygons (per pixel cluster): {len(gdf)}")

# ---------- 5. Build figure ----------
fig, axes = plt.subplots(2, 3, figsize=(18, 12))

def style(ax, title):
    ax.set_xlim(AOI["west"], AOI["east"])
    ax.set_ylim(AOI["south"], AOI["north"])
    ax.set_title(title, fontsize=11, fontweight="bold")
    ax.set_xlabel("Longitude (°W)", fontsize=9)
    ax.set_ylabel("Latitude (°N)", fontsize=9)
    ax.tick_params(labelsize=7)
    ax.set_aspect("equal")


def percentile_stretch(arr, p_lo=2, p_hi=98, gamma=0.85):
    """Linear stretch via p2/p98 percentiles, then gamma."""
    finite = arr[np.isfinite(arr)]
    if finite.size == 0:
        return np.zeros_like(arr)
    lo, hi = np.percentile(finite, [p_lo, p_hi])
    norm = np.clip((arr - lo) / max(hi - lo, eps), 0, 1)
    return np.power(norm, gamma)


# Panel A: true colour with p2/p98 stretch
ax = axes[0, 0]
rgb_true = np.dstack([
    percentile_stretch(bands["B4"]),
    percentile_stretch(bands["B3"]),
    percentile_stretch(bands["B2"]),
])
extent_s2 = [s2_bounds.left, s2_bounds.right, s2_bounds.bottom, s2_bounds.top]
ax.imshow(rgb_true, extent=extent_s2, origin="upper")
style(ax, "(a) Sentinel-2 true colour (B4-B3-B2)")
# Scale bar (~3 km at this latitude)
ax.plot([AOI["west"] + 0.005, AOI["west"] + 0.005 + 0.035],
        [AOI["south"] + 0.005] * 2, "k-", linewidth=2.3)
ax.text(AOI["west"] + 0.0225, AOI["south"] + 0.008, "~3 km",
        ha="center", fontsize=7, fontweight="bold", color="white",
        bbox=dict(facecolor="black", alpha=0.5, pad=1))
ax.annotate("N", xy=(0.95, 0.95), xycoords="axes fraction",
            fontsize=11, fontweight="bold", ha="center", va="top")
ax.annotate("", xy=(0.95, 0.95), xytext=(0.95, 0.85),
            xycoords="axes fraction",
            arrowprops=dict(arrowstyle="->", lw=1.3, color="black"))

# Panel B: SWIR false colour with p2/p98 stretch
ax = axes[0, 1]
rgb_swir = np.dstack([
    percentile_stretch(bands["B12"]),
    percentile_stretch(bands["B11"]),
    percentile_stretch(bands["B4"]),
])
ax.imshow(rgb_swir, extent=extent_s2, origin="upper")
style(ax, "(b) SWIR false colour (B12-B11-B4)")

# Panel C: Clay Ratio
ax = axes[0, 2]
p2, p98 = np.percentile(clay_ratio[np.isfinite(clay_ratio)], [2, 98])
im = ax.imshow(clay_ratio, extent=extent_s2, origin="upper",
               cmap="RdYlBu_r", vmin=p2, vmax=p98)
style(ax, "(c) Clay Ratio ($B_{11}/B_{12}$)")
plt.colorbar(im, ax=ax, shrink=0.7, pad=0.02, label="$B_{11}/B_{12}$")

# Panel D: SR per-pixel classification with 4 Cuprite-local formulas
ax = axes[1, 0]
cmap_class = ListedColormap([USGS_CLASS_COLORS[k] for k in sorted(USGS_CLASS_COLORS)])
norm_class = BoundaryNorm(np.arange(0.5, 5.5, 1), cmap_class.N)
ax.imshow(classmap, extent=extent_s2, origin="upper",
          cmap=cmap_class, norm=norm_class, interpolation="nearest")
style(ax, "(d) SR per-pixel classification (4 Cuprite formulas)")
patches = [Patch(facecolor=USGS_CLASS_COLORS[k], edgecolor="black",
                 label=USGS_CLASS_NAMES[k]) for k in sorted(USGS_CLASS_COLORS)]
ax.legend(handles=patches, loc="lower left", fontsize=7,
          framealpha=0.92, title="SR predicted class", title_fontsize=7.5)

# Panel E: USGS Rockwell GT raster
ax = axes[1, 1]
gt_display = np.where(gt > 0, gt, np.nan).astype(float)
extent_gt = [gt_bounds.left, gt_bounds.right, gt_bounds.bottom, gt_bounds.top]
ax.imshow(gt_display, extent=extent_gt, origin="upper",
          cmap=cmap_class, norm=norm_class, interpolation="nearest")
style(ax, "(e) USGS Rockwell (2017) ASTER-derived reference")
ax.legend(handles=patches, loc="lower left", fontsize=7,
          framealpha=0.92, title="USGS class", title_fontsize=7.5)

# Panel F: SR classification with USGS outlines overlaid
ax = axes[1, 2]
ax.imshow(classmap, extent=extent_s2, origin="upper",
          cmap=cmap_class, norm=norm_class, interpolation="nearest", alpha=0.85)
gdf.boundary.plot(ax=ax, edgecolor="black", linewidth=0.4, alpha=0.85)
style(ax, "(f) SR classification with USGS outlines overlaid")

fig.suptitle(
    "Cuprite, Nevada, USA — Sentinel-2 imagery, classical and SR indices, "
    "SR per-pixel classification, and USGS Rockwell (2017) reference comparison\n"
    "Median S2 composite Jan–Jun 2024, 30 m display resolution; "
    f"AOI {(AOI['east']-AOI['west'])*111*0.79:.0f} km × {(AOI['north']-AOI['south'])*111:.0f} km",
    fontsize=11, y=1.005,
)

plt.tight_layout()
out = FIG_DIR / "cuprite_comparison.png"
fig.savefig(out, dpi=600, bbox_inches="tight")
fig.savefig(FIG_DIR / "cuprite_comparison.pdf", bbox_inches="tight")
plt.close()
print(f"Saved: {out}")
